pkgs/development/compilers/dotnet/combine-packages-builder.sh

```diff
@@ -0,0 +1,36 @@
+source $stdenv/setup
+
+mkdir -p $out
+
+# CLI will follor real path, so we have to copy it
+cp $cli/dotnet $out/
+mkdir $out/bin
+ln -s $out/dotnet $out/bin/
+
+# Create runtime & host links
+for runtime in $runtimes; do
+for h in `ls $runtime/host`; do
+    mkdir -p $out/host/$h
+    ln -sf  $runtime/host/$h/* $out/host/$h
+done
+for i in `ls $runtime/shared`; do
+    mkdir -p $out/shared/$i
+    ln -sf $runtime/shared/$i/* $out/shared/$i
+done
+done
+
+# Link SDKs
+mkdir -p $out/{sdk,packs,templates,sdk/NuGetFallbackFolder}
+for sdk in $sdks; do
+    test -d $sdk/templates && ln -sf $sdk/templates/* $out/templates/
+    test -d $sdk/packs     && ln -sf $sdk/packs/* $out/packs/
+    test -d $sdk/sdk && for sd in `ls $sdk/sdk`; do
+        [ "$sd" = "NuGetFallbackFolder" ] && continue
+        ln -sf $sdk/sdk/$sd $out/sdk/
+    done
+    # Link NuGetFallbackFolder contents
+    test -d $sdk/sdk/NuGetFallbackFolder && find $sdk/sdk/NuGetFallbackFolder -type f -exec ln -sf {} $out/sdk/NuGetFallbackFolder/  \;
+done
+
+# Print installed components
+$out/bin/dotnet --info
```

Comment:
this is what I came up with:
```
source $stdenv/setup

# enumerate the child directories, and link their children
linkSubDirectories() {
    local sourceDir=$1
    local destDir=$2

    for h in `ls $sourceDir`; do
        mkdir -p $destDir/$h
        ln -sf  $sourceDir/$h/* $destDir/$h
    done
}

mkdir -p $out/bin

# CLI will follow real path, so we have to copy it
cp $cli/dotnet $out/
ln -s $out/dotnet $out/bin/

# Create runtime & host links
for runtime in $runtimes; do
    linkSubDirectories $runtime/host $out/host
    linkSubDirectories $runtime/shared $out/shared
done

# Link SDKs
mkdir -p $out/{sdk,host,shared,packs,templates,sdk/NuGetFallbackFolder}
for sdk in $sdks; do
    ln -sf $sdk/templates/* $out/templates/
    ln -sf $sdk/packs/* $out/packs/
    for sd in `ls $sdk/sdk`; do
        [ "$sd" = "NuGetFallbackFolder" ] && continue
        ln -sf $sdk/sdk/$sd $out/sdk/
    done
    linkSubDirectories $sdk/shared $out/shared
    linkSubDirectories $sdk/host $out/host
    # Link NuGetFallbackFolder contents
    test -d $sdk/sdk/NuGetFallbackFolder && find $sdk/sdk/NuGetFallbackFolder -type f -exec ln -sf {} $out/sdk/NuGetFallbackFolder/  \;
done

# Print installed components
$out/bin/dotnet --info
```

Reply:
The main difference is that it will prefer the contents of the sdk's host and shared directories, this gives a few benefits, mainly closure size, but also more runtimes available:

with revisions, an sdk and runtime of the same version won't increase closure size:
```
[08:35:41] jon@jon-desktop /home/jon/projects/nixpkgs ((ea6469760d3...))
$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ netcore_2_1 ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined
/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined	 747.2M

$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/
/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined	 747.2M
```

Reply:
vs the old behavior, it will fail to build without a runtime:

```
$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ netcore_2_1 ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.NETCore.App 2.1.13 [/nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined/
/nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined	 845.8M

$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ ]; cli= sdk_2_1; })"
these derivations will be built:
  /nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv
building '/nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv'...
A fatal error occurred, the folder [/nix/store/0pl7pkjx8220yw3h43zjdv5maisrjlp9-dotnet-core-combined/host/fxr] does not exist
builder for '/nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv' failed with exit code 131
```

Reply:
As mentioned further down, I have replaced the custom build script with a library function: `buildEnv`, which seams to work well.